contracts/interchain-token/src/contract.rs

```diff
@@ -20,27 +46,171 @@ impl InterchainToken {
+        write_metadata(&env, token_meta_data);
 
-        env.storage().instance().set(&DataKey::Owner, &owner);
+        env.storage().instance().set(&DataKey::Admin, &admin);
+        env.storage()
+            .persistent()
+            .set(&DataKey::Minter(minter), &true);
+        env.storage()
+            .persistent()
+            .set(&DataKey::Minter(interchain_token_service.clone()), &true);
+
+        env.storage().instance().set(
+            &DataKey::InterchainTokenServiceAddress,
+            &interchain_token_service,
+        );
 
         Ok(())
     }
 
-    pub fn owner(env: &Env) -> Result<Address, ContractError> {
+    pub fn token_id(env: &Env) -> Address {
+        env.storage().instance().get(&DataKey::TokenId).unwrap()
+    }
+
+    pub fn interchain_token_service(env: &Env) -> Address {
         env.storage()
             .instance()
-            .get(&DataKey::Owner)
-            .ok_or(ContractError::NotInitialized)
+            .get(&DataKey::InterchainTokenServiceAddress)
+            .unwrap()
     }
 
-    pub fn transfer_ownership(env: Env, new_owner: Address) -> Result<(), ContractError> {
-        let owner: Address = Self::owner(&env)?;
-        owner.require_auth();
+    pub fn mint(env: Env, minter: Address, to: Address, amount: i128) -> Result<(), ContractError> {
+        minter.require_auth_for_args((&to, amount).into_val(&env));
+
+        check_nonnegative_amount(amount);
 
-        env.storage().instance().set(&DataKey::Owner, &new_owner);
+        let admin = admin(&env);
+        if admin != minter {
+            let is_authorized = env
+                .storage()
+                .persistent()
+                .get::<_, bool>(&DataKey::Minter(minter))
+                .unwrap_or(false);
 
-        event::transfer_ownership(&env, owner, new_owner);
+            if !is_authorized {
+                return Err(ContractError::NotAuthorizedMinter);
+            }
+        }
+
+        extend_instance_ttl(&env);
+
+        receive_balance(&env, to.clone(), amount);
+        TokenUtils::new(&env).events().mint(admin, to, amount);
 
         Ok(())
     }
+
+    pub fn set_admin(env: Env, new_admin: Address) {
+        let admin = admin(&env);
+        admin.require_auth();
+
+        extend_instance_ttl(&env);
+
+        env.storage().instance().set(&DataKey::Admin, &admin);
+        TokenUtils::new(&env)
+            .events()
+            .set_admin(admin.clone(), new_admin.clone());
+
+        event::set_admin(&env, admin, new_admin);
+    }
+
+    pub fn add_minter(env: &Env, minter: Address) {
+        admin(env).require_auth();
+
+        env.storage()
+            .persistent()
+            .set(&DataKey::Minter(minter.clone()), &true);
+
+        event::add_minter(env, minter);
+    }
+}
+
+#[contractimpl]
+impl token::Interface for InterchainToken {
+    fn allowance(env: Env, from: Address, spender: Address) -> i128 {
+        extend_instance_ttl(&env);
```

Comment:
not sure if we need this for every method vs key methods like transfer, transfer_from?